### DISCUSSION TITLE
csvreader a new "core" csv reader library - Let's work together - Why? Why not? Thoughts?

### DISCUSSION THREAD

**Discussion by geraldb** [tilo/smarter_csv#208] [General]
Hello,
   Your smarter_csv library  and README was a great inspiration for using "plain" hashes for returning records in the new csvreader library (instead of CSV::Table, CSV::Row, etc. as done in the old std csv library).

  Anyways, as far as I can tell your library doesn't have its own csv parser and still uses the old std csv library under the hood. I invite you to consider to use the new modern csvreader (see <https://github.com/rubycocos/csvreader>) that offers more formats / dialects / variants, for example.

 Also since you're a leading CSV ruby expert and user it would be greatly appreciated if you could tell (don't hold any negatives back) what's missing in the new csvreader library and if it would be a good replacement - dare I say - for smarter_csv too (getting back in the long run to a new / better / modern ruby csv std base library)

Cheers. Prost.  Greetings from Vienna, Austria.

**Update Aug/2022**   -  Great to sea a response after 3+ years!  Note: I updated the link to the csvreader library now housed / living at the ruby code commons (cocos) org - with all (optional) add-on gems .

**Comment by tilo** [comment]
For now, I don't see a need to use a new CSV parser under the hood.
SmarterCSV just removed the dependency on the Ruby CSV library.

But I will have a look at your project and see if there are synergies to explore

**Reply by geraldb** [reply]
Great to see that SmarterCSV is back (after 3+ years or such)!   Anyways, I am happy to answer any questions - the csvreader (inspired by smarterCSV and others) now lives at  https://github.com/rubycocos/csvreader

> SmarterCSV just removed the dependency on the Ruby CSV library.

Sorry if I ask an obvious questions - if you removed the dependency on the ruby CSV library - what CSV parser do you use?  your own? 

Not sure what's your long term plan for smarterCSV  let me restate some years later today in 2022 the offer to work together on a better CSV library.   What is your motivation to keep going solo?  What's your "roadmap" / plan for smarterCSV?  As far as I remember a rewrite, that is, v2.0 is in the works - is that still valid in 2022?   Are you using the smarterCSV gem yourself "in production"? Where?

**Reply by geraldb** [reply]
Update  - Not really expecting an answer after all those years anyways.   I browsed your code
and as far as I can tell you are now using a custom c-extension for your parser
-> <https://github.com/tilo/smarter_csv/blob/main/ext/smarter_csv/smarter_csv.c>

If that is the parser than dare I say that is way too simplistic - the csv format is unfortunately more complex than splitting values by commas.   See <https://donatstudios.com/Falsehoods-Programmers-Believe-About-CSVs> or  from my humbel self <https://github.com/csvspecs>.   

If I may give you unasked advice after studying performance with csv parsing (or parsing in general) in ruby -  using a c-extension is overkill  (it's great for sure as an optional addon) but usually using a parser built with StringScanner gets you close to a c-extension speed (because StringScanner is a c-extension ;-).   The new CSV parser in the csv std library uses this approach as does hexapdf for the pdf parser and so on.

**Comment by tilo** [comment]
SmarterCSV has now it's own CSV reader implemented, including a C-extension for acceleration

**Comment by geraldb** [comment]
No worries Thanks for the update. I was trying to be friendly and helpful. Keep on going and all the best whatever you choose.
